### ISSUE TITLE
FR: Way to substract users, groups, tags from allow rules in ACL

### ISSUE THREAD

**Issue by rocket-science-ch** [tailscale/tailscale#5108]
### What are you trying to do?

In larger deployments with many users, defining access for 95% becomes a hassle without specifically naming every user in a group. Therefore, using autogroup:members is a good option. But what if you what everyone to be able to access a service, expect for these two external contractors (the 95% problem described above)?

### How should we solve this?

I suggest a new description (e.g. the ! sign) to actively substract these users, groups or tags from the larger set mentioned:

` "acls": [
{
"action": "accept",
"src":    [
"autogroup:members",
"!group:exclude_this_contractors"
],
"dst": [
"git:443,80,22",
"tag:external:*",
],
},`

**Comment by DentonGentry** [comment]
The intent with this is to synchronize group memberships with an external source that can apply whatever sophisticated membership criteria is desired.
- group sync for Okta is available now: https://tailscale.com/blog/sync-okta-groups/
- group sync for Active Directory is expected next: https://github.com/tailscale/tailscale/issues/979
- group sync for Google is planned, but harder as Google dropped support for the standard API for this function in 2019: https://github.com/tailscale/tailscale/issues/2861

**Comment by rocket-science-ch** [comment]
My original intent is to do this independently of external groups, purely using ACL defined entries.

**Comment by rocket-science-ch** [comment]
My original intent is to do this independently of external groups, purely using ACL defined entries.

**Comment by DentonGentry** [comment]
related: https://github.com/tailscale/tailscale/issues/5290

**Comment by ShakataGaNai** [comment]
This, or DENY ACL ( #6915 ), is fairly important. We have groups that may have access to a majority of systems, like DevOps. Their SRC/DST ACL is effectively */*:*. However, there may be small exceptions to that, such as systems/networks they shouldn't be able to access for security/compliance reasons.

Today the only answer is to move DevOps from */*:* to listing out EVERY network/host/tag manually, EXCEPT the perhaps one network in question. During first time getting started, that's not a big deal. But as usage with Tailscale grows, that becomes more problematic. Now every time a new network is added, someone needs to remember to update that allow list for DevOps. Eventually, it'll become huge, unwieldy, and very hard to audit.

Where as it is much simpler, much more secure (since it's clearer and easier to read), and much less prone to human error... if there is simply a way to say "DevOps can access everything but X". If that's boolean logic or DENY ACL, doesn't matter, but it's really big oversight to not have this. Almost every firewall and ACL system from Cisco to IPTables, has a DENY equivalent.

**Comment by vsarunas** [comment]
Another vote for being to exclude a particular group/tag from an ACL.

**Comment by chernogorsky** [comment]
I have about 60 routes in my route tables with a single aggreage /16 for the site
so in order to filter one network for a group I have to put 59 routes in allow, and make sure I never use filtered route anythere else.

so, yes, any / all of  access=deny, ![range] in src/dst, deny = [range] statement would be quite helpfull

**Comment by yqs112358** [comment]
I think this feature is becoming one of the main obstacles to the enterprise-scale use for Tailscale. The current absense of Deny-like expression will complicate permission assignment in large-scale system, in turn affects the willingness of potential users.

**Comment by jason-masson** [comment]
> Deny-like

Everyone has thier use cases but IMO the default deny, "zero trust", model isn't a bad thing it's a plus. You only need to craft what is allowed and then everything else is denied. 

The ability to craft complex allow rules is a problem though.  We need the ability to specify multiple criteria before the traffic is allow.  EG:  devices must have multiple tags before traffic is allowed. Also negate options would be nice. 😁

Keep up the good work!

**Comment by yqs112358** [comment]
> > Deny-like
> 
> Everyone has thier use cases but IMO the default deny, "zero trust", model isn't a bad thing it's a plus. You only need to craft what is allowed and then everything else is denied.
> 
> The ability to craft complex allow rules is a problem though. We need the ability to specify multiple criteria before the traffic is allow. EG: devices must have multiple tags before traffic is allowed. Also negate options would be nice. 😁
> 
> Keep up the good work!

Sure, you are right. So I believe that devs will prefer Boolean and Combinatorial grammers too, which provides easier node/tag selection while maintaining the zero-turst model. 
This can be equally effective in addressing those issues. 😃

**Comment by jason-masson** [comment]
> > > Deny-like
> > 
> Sure, you are right. So I believe that devs will prefer Boolean and Combinatorial grammers too, which provides easier node/tag selection while maintaining the zero-turst model. 
> This can be equally effective in addressing those issues. 😃


Yeah. We are saying similar things.   At the core, we need the ability to craft more complex policies to drive adoption.

**Comment by vsarunas** [comment]
Will negate be part of [Grants](https://tailscale.com/kb/1538/grants-syntax) in the future? In order to create rules which allow everyone except a restricted group?

**Comment by tonhe** [comment]
+1 for this.

The scenario a deny action solves cleanly: excluding a subnet from a broader allowed supernet on the destination side. Allow a /16, deny a /24 inside it. Today there's no way to express that, so you end up allowing a series of hand compused CIDR ranges that exclude the ranges you need to "deny"...  and re-deriving it whenever that deny range changes. A deny targeting a dst CIDR collapses that to one additional line.

**Comment by timtailscale** [comment]
@tonhe the IP set feature already allows you to solve the issue you describe for subnet routes: [https://tailscale.com/docs/features/tailnet-policy-file/ip-sets](https://tailscale.com/docs/features/tailnet-policy-file/ip-sets).
